pkg/ipcache/ipcache.go

```diff
@@ -180,77 +189,193 @@ func allowOverwrite(existing, new Source) bool {
+			listener.OnIPIdentityCacheChange(Upsert, *cidr, hostIP, oldIdentity, newIdentity.ID)
+		}
 	}
 
 	return true
 }
 
 // deleteLocked removes removes the provided IP-to-security-identity mapping
 // from ipc with the assumption that the IPCache's mutex is held.
-func (ipc *IPCache) deleteLocked(IP string) {
-	log.WithFields(logrus.Fields{
-		logfields.IPAddr: IP,
-	}).Debug("Removing from ipcache layer")
+func (ipc *IPCache) deleteLocked(ip string) {
+	scopedLog := log.WithFields(logrus.Fields{
+		logfields.IPAddr: ip,
+	})
 
-	identity, found := ipc.ipToIdentityCache[IP]
-	if found {
-		delete(ipc.ipToIdentityCache, IP)
-		delete(ipc.identityToIPCache[identity.ID], IP)
-		if len(ipc.identityToIPCache[identity.ID]) == 0 {
-			delete(ipc.identityToIPCache, identity.ID)
-		}
+	cachedIdentity, found := ipc.ipToIdentityCache[ip]
+	if !found {
+		scopedLog.Debug("Attempt to remove non-existing IP from ipcache layer")
+		return
+	}
+
+	var cidr *net.IPNet
+	cacheModification := Delete
+	var oldIdentity *identity.NumericIdentity
+	newIdentity := cachedIdentity
+	callbackListeners := true
 
+	var err error
+	if _, cidr, err = net.ParseCIDR(ip); err == nil {
 		// Remove a reference for the prefix length if this is a CIDR.
-		if _, cidr, err := net.ParseCIDR(IP); err == nil {
-			pl, bits := cidr.Mask.Size()
-			switch bits {
-			case net.IPv6len * 8:
-				unrefPrefixLength(ipc.v6PrefixLengths, pl)
-			case net.IPv4len * 8:
-				unrefPrefixLength(ipc.v4PrefixLengths, pl)
+		pl, bits := cidr.Mask.Size()
+		switch bits {
+		case net.IPv6len * 8:
+			unrefPrefixLength(ipc.v6PrefixLengths, pl)
+		case net.IPv4len * 8:
+			unrefPrefixLength(ipc.v4PrefixLengths, pl)
+		}
+
+		// Check whether the deleted CIDR was shadowed by an endpoint IP. In
+		// this case, skip calling back the listeners since they don't know
+		// about its mapping.
+		if _, endpointIPFound := ipc.ipToIdentityCache[cidr.IP.String()]; endpointIPFound {
+			scopedLog.Info("Deleting CIDR shadowed by endpoint IP")
+			callbackListeners = false
+		}
+	} else if endpointIP := net.ParseIP(ip); endpointIP != nil { // Endpoint IP.
+		// Convert the endpoint IP into an equivalent full CIDR.
+		bits := net.IPv6len * 8
+		if endpointIP.To4() != nil {
+			bits = net.IPv4len * 8
+		}
+		cidr = &net.IPNet{
+			IP:   endpointIP,
+			Mask: net.CIDRMask(bits, bits),
+		}
+
+		// Check whether the deleted endpoint IP was shadowing that CIDR, and
+		// restore its mapping with the listeners if that was the case.
+		if cidrIdentity, cidrFound := ipc.ipToIdentityCache[cidr.String()]; cidrFound {
+			if cidrIdentity.ID != cachedIdentity.ID {
+				scopedLog.Info("Deleting endpoint IP shadowing CIDR, restoring CIDR")
+				cacheModification = Upsert
+				oldIdentity = &cachedIdentity.ID
+				newIdentity = cidrIdentity
+			} else {
+				// The endpoint IP and the CIDR were associated with the same
+				// identity. Nothing changes for the listeners.
+				callbackListeners = false
```

Comment:
This needs to take into account a potential change in the hostIP as well. Otherwise we may continue to encap.

Reply:
Yes, agreed. :/
I think this will require keeping a mapping of IP to host IP, so we know whether the host IP changes or not because of un/shadowing.
I could also remove this optimization here and add an equivalent check into the NPHDS callback to return immediately if the IDs are identical.
Any preferences?

Reply:
Should hostIP changes be a separate cache modification type? Then the proxy can ignore it easily. Only the BPF side needs to be aware of such changes, right?

Reply:
I’ve been thinking about this. I would like that very much. However I don’t think that would be possible since the ID and host IP need to change together to be able to compute the correct BPF map key, cf the callback implementation in the daemon.